velox/functions/sparksql/URLFunctions.h

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/regex.hpp>
+#include <folly/CPortability.h>
+#include <cctype>
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+/// Performs initial validation of the URI.
+/// Checks if the URI contains ascii whitespaces or
+/// unescaped '%' chars.
+bool isValidURI(StringView input) {
+  const char* p = input.data();
+  const char* end = p + input.size();
+  char buf[3];
+  buf[2] = '\0';
+  char* endptr;
+  for (; p < end; ++p) {
+    if (facebook::velox::functions::stringImpl::isAsciiWhiteSpace(*p)) {
+      return false;
+    }
+
+    if (*p == '%') {
+      if (p + 2 < end) {
+        buf[0] = p[1];
+        buf[1] = p[2];
+        strtol(buf, &endptr, 16);
+        p += 2;
+        if (endptr != buf + 2) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+static const boost::regex kUriRegex(
+    "(([a-zA-Z][a-zA-Z0-9+.-]*):)?" // scheme:
+    "([^?#]*)" // authority and path
+    "(?:\\?([^#]*))?" // ?query
+    "(?:#(.*))?"); // #fragment
+template <typename T>
+T submatch(const boost::cmatch& match, int indx);
+
+template <>
+FOLLY_ALWAYS_INLINE std::string_view submatch(
+    const boost::cmatch& match,
+    int idx) {
+  const auto& sub = match[idx];
+  return std::string_view(sub.first, sub.length());
+}
+
+template <>
+FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
+  const auto& sub = match[idx];
+  return StringView(sub.first, sub.length());
+}
+
+template <typename TInString>
+bool parse(const TInString& rawUrl, boost::cmatch& match) {
+  if (!isValidURI(rawUrl)) {
+    return false;
+  }
+  return boost::regex_match(
+      rawUrl.data(), rawUrl.data() + rawUrl.size(), match, kUriRegex);
+}
+// PARSE_URL(url, partToExtract) → string
+// PARSE_URL(url, partToExtract, key) → string
+//
+// Extracts a part of a URL. The partToExtract argument can be one of
+// 'PROTOCOL', 'HOST', 'PATH', 'REF', 'AUTHORITY', 'FILE', 'USERINFO',
```

Comment:
Do these have to be in all caps? Would be nice to clarify in the docs.

Reply:
Yes, they do have to be all capitalized. Below is output from my sanity checker confirming this:

<img width="191" alt="image" src="https://github.com/facebookincubator/velox/assets/14911410/4a92967e-f8ad-4c24-847b-bc229aad4fae">

Will update documentation to make this more clear